Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/saf/file/AbstractFile.kt

```diff
@@ -0,0 +1,315 @@
+ *      .appendSubDirSegment("dir2")
+ *      .appendSubDirSegment("dir3")
+ *      .appendFileNameSegment("test.txt")
+ *      .createNew();
+ *
+ *  Then you can call methods that are similar to the standard Java File API, e.g. [exists],
+ *  [getName], [getLength], [isFile], [isDirectory], [canRead], [canWrite] etc.
+ *
+ *  If you want to work with multiple files in a directory (or sub directories) you may want
+ *  to [clone] the file that represents that directory, e.g:
+ *
+ *  AbstractFile clonedFile = baseDir.clone();
+ *
+ *  AbstractFile f1 = clonedFile
+ *      .appendFileNameSegment("f1.txt")
+ *      .createNew();
+ *  AbstractFile f2 = clonedFile
+ *      .appendFileNameSegment("f2.txt")
+ *      .createNew();
+ *  AbstractFile f3 = clonedFile
+ *      .appendFileNameSegment("f3.txt")
+ *      .createNew();
+ *
+ *  You have to do this because some methods may mutate the internal state of the [AbstractFile], so
+ *  after calling, let's say:
+ *
+ *  AbstractFile f1 = baseDir
+ *      .appendFileNameSegment("f1.txt")
+ *      .createNew();
+ *
+ *  Without cloning it first baseDir will be start pointing to <Kuroba_Base_Dir>/f1.txt instead of
+ *  just <Kuroba_Base_Dir>. The same thing applies to any method marked with [MutableMethod] annotation.
+ *  methods marked with [ImmutableMethod] do this stuff internally so they are safe to use without
+ *  cloning.
+ *
+ *  Sometimes you don't know which external directory to choose to store a new file (the SAF or the
+ *  old raw Java File external directory). In this case you can use:
+ *
+ *  AbstractFile baseDir = fileManager.newSaveLocationFile();
+ *
+ *  Method which will create an [AbstractFile] with root pointing to either Kuroba SAF base directory
+ *  (if user has set it) or if he didn't then to the default external directory (Backed by raw
+ *  Java File) or
+ *
+ *  AbstractFile baseDir = fileManager.newLocalThreadFile();
+ *
+ *  Method which will create an [AbstractFile] with root pointing to either user's selected local
+ *  threads directory or to the default external local threads directory
+ *
+ * */
+abstract class AbstractFile(
+        /**
+         * /test/123/test2/filename.txt -> 4 segments
+         * */
+        protected val segments: MutableList<Segment>
+) {
+    /**
+     * Appends a new subdirectory (or couple of subdirectories, e.g. "dir1/dir2/dir3")
+     * to the root directory
+     * */
+    @MutableMethod
+    abstract fun appendSubDirSegment(name: String): AbstractFile
+
+    /**
+     * Appends a file name to the root directory (or couple subdirectories with filename at the end,
+     * e.g. "dir1/dir2/dir3/test.txt"
+     * */
+    @MutableMethod
+    abstract fun appendFileNameSegment(name: String): AbstractFile
+
+    /**
+     * Creates a new file that consists of the root directory and segments (sub dirs or the file name)
+     * Behave similarly to Java's mkdirs() method but work not only with directories but files as well.
+     * */
```

Comment:
Add: So if you have a few segments, lets say, "dir1/dir2/dir3/test.txt" then all of the subdirs and the file will be created on the disk when calling this method. If some error occurs while trying to create a subdir or a file then already created directories will remain intact (you have to delete them manually).